test/e2e/self_hosted.go

```diff
@@ -73,17 +90,71 @@ func SelfHostedSpec(ctx context.Context, inputGetter func() SelfHostedSpecInput)
 		Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling %s spec", specName)
 		Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
 		Expect(os.MkdirAll(input.ArtifactFolder, 0750)).To(Succeed(), "Invalid argument. input.ArtifactFolder can't be created for %s spec", specName)
-		Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersion))
+
+		if !input.EnableUpgrade {
+			// Use KubernetesVersion if no upgrade step is defined by test input.
+			Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersion))
+
+			kubernetesVersion = input.E2EConfig.GetVariable(KubernetesVersion)
+		} else {
+			Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersionUpgradeFrom))
+			Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersionUpgradeTo))
+			Expect(input.E2EConfig.Variables).To(HaveKey(EtcdVersionUpgradeTo))
+			Expect(input.E2EConfig.Variables).To(HaveKey(CoreDNSVersionUpgradeTo))
+
+			kubernetesVersion = input.E2EConfig.GetVariable(KubernetesVersionUpgradeFrom)
+		}
 
 		// Setup a Namespace where to host objects for this spec and create a watcher for the namespace events.
 		namespace, cancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder)
 		clusterResources = new(clusterctl.ApplyClusterTemplateAndWaitResult)
+
+		if input.ControlPlaneMachineCount == nil {
+			controlPlaneMachineCount = 1
+		} else {
+			controlPlaneMachineCount = *input.ControlPlaneMachineCount
+		}
+
+		if input.WorkerMachineCount == nil {
+			workerMachineCount = 1
+		} else {
+			workerMachineCount = *input.WorkerMachineCount
+		}
 	})
 
 	It("Should pivot the bootstrap cluster to a self-hosted cluster", func() {
 		By("Creating a workload cluster")
 
 		workloadClusterName := fmt.Sprintf("%s-%s", specName, util.RandomString(6))
+		variableOverwrites := map[string]string{}
+
+		isDockerInfrastructureProvider := false
```

Comment:
Can we move this detection in a util func at the end of the file?

Reply:
Hm that the CAPD provider is deployed doesn't necessarily mean that this test is using a cluster-template using CAPD
If I understand it correctly the e2e framework can deploy multiple infra providers.

Given that, I wonder if we should just always set those variables (could be fine, given that they are namespaced to DOCKER)

Reply:
Maybe it would be more implicit by using `CAPD_` instead of `DOCKER_` prefix?

Reply:
I was thinking about that as well. Then I looked at the quickstart and noticed that basically all provider env vars are just using the "cloud" as prefix: AWS_, AZURE_, ...

Reply:
ack.

moved to seperate func.